README.md
# values.js

[![NPM Version][npm-image]][npm-url]
[![Minzipped size][bundlephobia-image]][bundlephobic-url]
[![License][license-image]][license-url]
![][github-actions-nodejs-image]
[![codecov][codecov-image]][codecov-url]
[![Known Vulnerabilities][snyk-image]][snyk-url]
<!--[![Dependencies][david-image]][david-url]
[![Dev Dependencies][david-dev-image]][david-dev-url]-->
[![Libraries.io dependency status for latest release][librariesio-img]][librariesio-url]
[![Total alerts][lgtm-image]][lgtm-url]
[![Language grade: JavaScript][lgtm-grade-image]][lgtm-grade-url]

Get tints and shades of a CSS color

> _The lightness or darkness of a color is called its value.
Tints are light values that are made by mixing a color with white, which increases lightness. Shades are dark values that are made by mixing a color with black, which reduces lightness._

https://noeldelgado.github.io/values.js/

### Supports
* \<color value\>
	* Hexadecimal RGB value: #RGB #RRGGBB
	* #RGBA #RRGGBBAA (4 and 8-digit hexadecimal RGBA notation)
	* RGB/A - CSS Color Module Level 3 and 4 (number, percentage)
	* HSL/A - CSS Color Module Level 3 and 4 (number, deg, rad, turn)
* \<color keyword\>
	* One of the [pre-defined color keywords](https://www.w3.org/wiki/CSS/Properties/color/keywords).
* transparent
	* Shorthand for transparent black, rgba(0,0,0,0).

## Installation

**NPM**

```sh
npm install values.js --save
```

Or as a `<script>` tag from a CDN as `Values`:

**Unpkg CDN**

```html
<script src="https://unpkg.com/values.js"></script>
```

**jsDelivr CDN**

```html
<script src="https://cdn.jsdelivr.net/npm/values.js"></script>
```

## Usage
```js
import Values from 'values.js'
const color = new Values('hsl(204deg 100% 50% / 1)'), { log } = console

log(color.tint(25))
//> { rgb: [64, 179, 255], alpha: 1, type: "tint", weight: 25, ...methods }
log(color.shade(12))
//> { rgb: [0, 135, 224], alpha: 1, type: "shade", weight: 12, ...methods }
log(color.tints(8))
//> (12) [Values...]
log(color.shades(23))
//> (4) [Values...]
log(color.all(20))
//> (11) [Values...]

// instance example for 'red'
Values {
  alpha: 1
  rgb: (3) [255, 0, 0]
  type: "base"
  weight: 0
  get hex: ƒ(...)
  setColor: ƒ setColor(color)
  tint: ƒ tint(weight=50)
  tints: ƒ tints(weight=10)
  shade: ƒ shade(weight=50)
  shades: ƒ shades(weight=10)
  all: ƒ all(weight=10)
  hexString: ƒ hexString()
  rgbString: ƒ rgbString()
  getBrightness: ƒ getBrightness()
}
```
See [tests](https://github.com/noeldelgado/values.js/tree/master/test) for more cases.

## API

### constructor(color)
Throws if the color is not accepted.

- `@param {string} color` — a valid CSS color string

### setColor(color)
Sets a new base color, returns `null` if color has not been accepted.

- `@param {string} color` - a valid CSS color string
- `@return {Values|null}`

### tint([weight=50])
Lightens the base color by mixing it with white as specified by weight.

- `@param {number} [weight=50]`
- `@return {Values}`

### shade([weight=50)
Darkens the base color by mixing it with black as specified by weight.

- `@param {number} [weight=50]`
- `@return {Values}`

### tints([weight=10])
Generates the tints of the base color as specified by weight.

- `@param {number} [weight=10]`
- `@return {Array<Values>}`

### shades([percentage=10])
Generates the shades of the base color as specified by weight.

- `@param {number} [weight=10]`
- `@return {Array<Values>}`

### all([weight=10])
Generates the tints and shades of the base color as specified by weight.

- `@param {number} [weight=10]`
- `@return {Array<Values>}`

### hexString()
Returns the color in hexadecimal RGB notation.

- `@returns {string}` @example `#000000` or `#00000080`

### rgbString()
Returns the color in rgb() functional notation.

- `@returns {string}` @example `rgb(0, 0, 0)` or `rgba(0, 0, 0, 0.5)`

### getBrightness()
Calculates the brightness of the color.

- `@return {number}` — the base-color brightness.

## Dev
```sh
npm install 	# install dev-dependencies
npm test		# run the tests
npm run dev 	# watch for changes and run tests
```

## Related
- [shadowlord](https://github.com/noeldelgado/shadowlord) - Tints and shades generator tool
- [mix-css-color](https://github.com/noeldelgado/mix-css-color) - Mix two CSS colors together in variable proportion. Opacity is included in the calculations.
- [parse-css-color](https://github.com/noeldelgado/parse-css-color) - Parse a CSS color string

## License
MIT © [Noel Delgado](https://pixelia.me/)

[npm-image]: https://img.shields.io/npm/v/values.js.svg?logo=npm&label=NPM
[npm-url]: https://www.npmjs.com/package/values.js
[github-actions-nodejs-image]: https://github.com/noeldelgado/values.js/workflows/Node.js%20CI/badge.svg
[github-actions-lighthouse-image]: https://github.com/noeldelgado/values.js/workflows/Lighthouse/badge.svg
[codecov-image]: https://codecov.io/gh/noeldelgado/values.js/branch/master/graph/badge.svg
[codecov-url]: https://codecov.io/gh/noeldelgado/values.js
[license-image]: https://img.shields.io/npm/l/values.js.svg
[license-url]: https://github.com/noeldelgado/values.js/blob/master/LICENSE
[bundlephobia-image]: https://img.shields.io/bundlephobia/minzip/values.js?label=size
[bundlephobic-url]: https://bundlephobia.com/result?p=values.js
[snyk-image]: https://snyk.io/test/npm/values.js/badge.svg
[snyk-url]: https://snyk.io/test/npm/values.js
[david-image]: https://img.shields.io/david/noeldelgado/values.js.svg
[david-url]: https://david-dm.org/noeldelgado/values.js
[david-dev-image]: https://img.shields.io/david/dev/noeldelgado/values.js.svg
[david-dev-url]: https://david-dm.org/noeldelgado/values.js?type=dev
[librariesio-img]: https://img.shields.io/librariesio/release/npm/values.js
[librariesio-url]: https://libraries.io/npm/values.js
[lgtm-image]: https://img.shields.io/lgtm/alerts/g/noeldelgado/values.js.svg?logo=lgtm&logoWidth=18
[lgtm-url]: https://lgtm.com/projects/g/noeldelgado/values.js/alerts/
[lgtm-grade-image]: https://img.shields.io/lgtm/grade/javascript/g/noeldelgado/values.js.svg?logo=lgtm&logoWidth=18
[lgtm-grade-url]: https://lgtm.com/projects/g/noeldelgado/values.js/context:javascript


CHANGELOG.md
# Changelog

## [Unreleased]

## [2.1.1] - [#24](https://github.com/noeldelgado/values.js/pull/24) - 2022-03-30
### Fixed
- types: `values.rgb`

## [2.1.0] - [#21](https://github.com/noeldelgado/values.js/pull/21) - 2022-03-23
### Added
- TS type definitions
### Changed
- README: replace david-dm.org with libraries.io
- upgrade and pin dependencies:
  - mix-css-color 0.2.0
  - parse-css-color 0.2.0
  - parse-color 1.3.0

## [2.0.0] - [#7](https://github.com/noeldelgado/values.js/pull/7) - 2020-05-23
### Added
- alpha support
  - #RRGGBBAA
  - RGB CSS Color Module Level 4
  - HSL CSS Color Module Level 4 (number, deg, rad & turn)
  - `alpha` property to hold a value within `0` and `1`
- [pre-defined color keywords](https://www.w3.org/wiki/CSS/Properties/color/keywords) support
- transparent - Shorthand for transparent black, rgba(0,0,0,0)
- default value `#000` for when `undefined`, `null`  or `<empty>` is passed to the constructor
- editorconfig
- .eslintrc
- `lighthouse-ci` action for `gh-pages`

### Changed
- refactor code
  - move validations to [parse-css-color](https://github.com/noeldelgado/parse-css-color)
  - move mix color functionality to [mix-css-color](https://github.com/noeldelgado/mix-css-color)
  - add [pure-color](https://github.com/WickyNilliams/pure-color) for color format conversion
- `throw` error if input is not a CSS string recognized by `parse-css-color`
- `rgb` property now returns an array in the form of `[r, g, b]` instead of `{ r: number, g: number, b: number }`
- `hex` is now a getter (access backward-compatible)
  - if alpha is less than `1` it returns 8 digits now `#RRGGBBAA` otherwise 6 `#RRGGBB`
- `hexString` method return value could include alpha
  - if alpha is less than `1` it returns 8 digits now `#RRGGBBAA` otherwise 6 `#RRGGBB`
- `isTint`, `isShade` and `isBaseColor` properties replaced by new `type` field
  - `type="tint"`
  - `type="shade"`
  - `type="base"`
- `percentage` property is now `weight` and, it is included on every instance instead of just for tints and shades
- `setColor` return null if color is not accepted

### Removed
- bower distribution
- properties
  - `hsl`
  - `isTint`, `isShade`, `isBaseColor`
- methods
  - `hslString`
- static utils functions
  - `isHex`, `isRGB` and `isHSL` logic moved to [parse-css-color](https://github.com/noeldelgado/parse-css-color)
  - `mix` (moved to [mix-css-color](https://github.com/noeldelgado/mix-css-color))

## [1.1.1] - [#6](https://github.com/noeldelgado/values.js/pull/6) - 2020-04-16
### Fixed
- shade is not defined

## [1.1.0] - [#5](https://github.com/noeldelgado/values.js/pull/5) - 2020-03-10
### Added
- ✨ new properties to the Array result of `vallues.all()` [`5f204a2`](https://github.com/noeldelgado/values.js/commit/5f204a2b4757bf0e2e77910ae50f44bbb19af56f)
  - `isTint: true` for tints
  - `isShade: true` for shades
  - `isBaseColor: true` for the original input color
- Added Changelog (this)

### Changed
- updated README [`f226453`](https://github.com/noeldelgado/values.js/commit/5f204a2b4757bf0e2e77910ae50f44bbb19af56f)
  - added params signature for methods
  - added `hexString`, `rgbString` and `hslString` missing methods
- updated tests for new properties

## [1.0.3] - 2015-10-20
### Added
- npm and bower version badges [`a7f04f5`](https://github.com/noeldelgado/values.js/commit/a7f04f5f1b55580a8d5c3583648bfdb398c0b5bd)

### Fixed
- #3, {tint,shade} === 0 returning default of 50% [`8d83fa0`](https://github.com/noeldelgado/values.js/commit/8d83fa003a6011352034e6d858afcc55e6009e2d)
- jshint errors [`3f54dc6`](https://github.com/noeldelgado/values.js/commit/3f54dc697814582514063e1feda9053d97b76eae)

## [1.0.2] - 2015-03-12
### Fixed
- CommonJS on linux: Require() is case sensitive on linux, so the example only worked on OSX. [#1](https://github.com/noeldelgado/values.js/pull/1)

## [1.0.1] - 2015-03-08
### Changed
- white list distributed files on NPM [`fc2ece8`](https://github.com/noeldelgado/values.js/commit/fc2ece8da029a3d8b27e910371342400542a1bc8)

### Fixed
- revert / return duplicated tints and shades, using an object to filter was messing the results ordering [`79294f8`](https://github.com/noeldelgado/values.js/commit/79294f81dba7fbe1358a7ba41bfa2b91a78c6eb1)

## [1.0.0] - 2015-03-08
### Changed
- refactor [`55bc6e6`](https://github.com/noeldelgado/values.js/commit/55bc6e69d5dccfcd73178f990a2b87636f8d34fc)
  - fixed tints and shades output actually mixing with white and black
- renamed methods
  - getTints => tints
  - getShades => shades
  - getAll => all

### Added
- methods
  - tint
  - shade
  - getBrightness
  - hexString
  - rgbString
  - hslString

### Removed
- methods
  - setStep
  - lighteness
- minified dist version

## [0.1.5] - 2014-09-16
### Fixed
- ignore case to isRGB and isHSL Util methods [`17cf7d5`](https://github.com/noeldelgado/values.js/commit/17cf7d55b105db314a623eb18c7d3349ab939f68)

## [0.1.4] - 2014-09-16
- 0.1.4 [`287ba66`](https://github.com/noeldelgado/values.js/commit/287ba6654c327d84debac31f30b77e73cb4aaaed)
  - add / hsl validaton support
  - update / rgb validation
  - add / tests

## [0.1.3] - 2014-09-06
## [0.1.2] - 2014-07-12
### Added
- bower support

### Fixed
- NPM main file

## [0.1.1] - 2014-07-02
- initial release

[Unreleased]: https://github.com/noeldelgado/values.js/compare/v2.1.0...HEAD
[2.1.1]: https://github.com/noeldelgado/values.js/compare/v2.1.0...v2.1.1
[2.1.0]: https://github.com/noeldelgado/values.js/compare/v2.0.0...v2.1.0
[2.0.0]: https://github.com/noeldelgado/values.js/compare/v1.1.1...v2.0.0
[1.1.1]: https://github.com/noeldelgado/values.js/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/noeldelgado/values.js/compare/v1.0.3...v1.1.0
[1.0.3]: https://github.com/noeldelgado/values.js/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/noeldelgado/values.js/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/noeldelgado/values.js/compare/v1.0.0...v1.0.1

source/values.js
(function () {

    "use strict";

    var Values = function Values(options) {
        this.hex            = '#000000';
        this.rgb            = 'rgb(0, 0, 0)';
        this.rgba           = 'rgba(0, 0, 0, 0)';
        this.hsl            = 'hsl(0, 0%, 0%)';
        this.hsla           = 'hsla(0, 0%, 0%, 0)';
        this.brightness     = 0;
        this.step           = 1;
        this.tints          = [];
        this.shades         = [];
        this.all            = [];
        this._rgb           = {r: 0, g: 0, b: 0};
        this._rgba          = {r: 0, g: 0, b: 0, a: 0};
        this._hsl           = {h: 0, s: 0, l: 0};
        this._hsla          = {h: 0, s: 0, l: 0, a: 0};

        if (typeof options === "string") {
            this.setColor(options);
        }
    };

    Values.Utils = {
        _reHEX : new RegExp("^(#)?([0-9a-fA-F]{3})([0-9a-fA-F]{3})?$"),
        _reRGB : new RegExp("rgba?\\s*\\((\\d+)\\,\\s*(\\d+)\\,\\s*(\\d+)(,\\s*(\\d+(\\.\\d+)?))?\\)","i"),
        _reHSL : new RegExp("hsla?\\((\\d+),\\s*([\\d.]+)%,\\s*([\\d.]+)%,?\\s*(?:0(\\.\\d+)?|1(\\.0)?)?\\)","i"),

        isHEX : function isHEX(value) {
            return this._reHEX.test(value);
        },
        isRGB : function isRGB(value) {
            var result, rgb, r, g, b;

            result = false;

            if (this._reRGB.test(value)) {
                rgb = value.match(this._reRGB);
                r = rgb[1];
                g = rgb[2];
                b = rgb[3];

                if ((r <= 255) && (g <= 255) && (b <= 255)) {
                    result = true;
                }
            }

            return result;
        },
        isHSL : function isHSL(value) {
            var result, hsl, h, s, l;

            result = false;

            if (this._reHSL.test(value)) {
                hsl = value.match(this._reHSL);
                h = hsl[1];
                s = hsl[2];
                l = hsl[3];

                if ((h <= 360) && (s <= 100) && (l <= 100)) {
                    result = true;
                }
            }

            return result;
        },

        RGBA : function RGBA(color, alpha) {
            var rgb = HEXtoRGB(color);
            return 'rgba('+rgb.r+', '+rgb.g+', '+rgb.b+', '+alpha+')';
        }
    };

    var HEXtoRGB = function (color) {
        var r, g, b;
        color = color.replace('#', '');
        if ( color.length === 6 ) {
            r = parseInt(color.slice(0,2), 16);
            g = parseInt(color.slice(2,4), 16);
            b = parseInt(color.slice(4,6), 16);
        } else {
            r = parseInt(color.slice(0,1) + color.slice(0,1), 16);
            g = parseInt(color.slice(1,2) + color.slice(1,2), 16);
            b = parseInt(color.slice(2,3) + color.slice(2,3), 16);
        }
        return {r: r, g: g, b: b};
    };

    var RGBtoHEX = function (r, g, b) {
        r = parseInt(r).toString(16);
        g = parseInt(g).toString(16);
        b = parseInt(b).toString(16);
        r = r.length === 1 ? "0" + r : r;
        g = g.length === 1 ? "0" + g : g;
        b = b.length === 1 ? "0" + b : b;
        return '#' + r + '' + g + '' + b;
    };

    var RGBtoHSL = function(r, g, b) {
        var r = r/255,
            g = g/255,
            b = b/255,
            max = Math.max(r, g, b),
            min = Math.min(r, g, b),
            h, s, l = (max + min) / 2;
        if (max === min) {
            h = s = 0;
        } else {
            var d = max - min;
            s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
            switch(max) {
                case r: h = (g - b) / d + (g < b ? 6 : 0); break;
                case g: h = (b - r) / d + 2; break;
                case b: h = (r - g) / d + 4; break;
            }
            h /= 6;
        }
        return {h: h, s: s, l: l};
    };

    var HUEtoRGB = function (v1, v2, vh) {
        if (vh < 0) vh += 1;
        if (vh > 1) vh -= 1;
        if ((6 * vh) < 1) return (v1 + (v2 - v1) * 6 * vh);
        if ((2 * vh) < 1) return v2;
        if ((3 * vh) < 2) return (v1 + (v2 - v1) * ((2/3) - vh) * 6);
        return v1;
    }

    var HSLtoRGB = function(h, s, l) {
        var r, g, b;
        if (s == 0) {
            r = g = b = l;
        } else {
            var q = l < 0.5 ? l * (1 + s) : l + s - l * s;
            var p = 2 * l - q;
            r = HUEtoRGB(p, q, h + 1/3);
            g = HUEtoRGB(p, q, h);
            b = HUEtoRGB(p, q, h - 1/3);
        }

        return {r: r * 255, g: g * 255, b: b * 255};
    };

    var getBrightness = function getBrightness(rgb) {
        var sum = rgb.r + rgb.g + rgb.b;
        return Math.round(sum / (255 * 3) * 100);
    };

    var __setValues = function __setValues(obj, rgb, h, s, l) {
        var r = Math.round(rgb.r),
            g = Math.round(rgb.g),
            b = Math.round(rgb.b);

        obj._rgb    = {r: r, g: g, b: b};
        obj._hsl    = {h: h, s: s, l: l};
        obj.hex     = RGBtoHEX(r, g, b);
        obj._rgba   = obj._rgb.a = 1;
        obj._hsla   = obj._hsl.a = 1;
        obj.rgb     = 'rgb('+r+', '+g+', '+b+')';
        obj.rgba    = 'rgba('+r+', '+g+', '+b+', 1)';
        obj.hsl     = 'hsl('+h+', '+s+'%, '+l+'%)';
        obj.hsla    = 'hsla('+h+', '+s+'%, '+l+'%, 1)';
        obj.brightness = getBrightness(obj._rgb);

        return obj;
    };

    Values.prototype.setColor = function setColor(value) {
        var $hex, $rgb, $hsl, $h, $s, $l, _rgb, _hsl

        if (Values.Utils.isHEX(value)) {
            if (!/^#/.test(value)) {
                value = '#' + value;
            }

            $hex = value;
        } else if (Values.Utils.isRGB(value)) {
            _rgb = value.replace(/[^\d,]/g, '').split(',');

            $hex = RGBtoHEX( _rgb[0], _rgb[1], _rgb[2] );
        } else if (Values.Utils.isHSL(value)) {
            _hsl = value.match(Values.Utils._reHSL);
            _rgb = HSLtoRGB(_hsl[1]/360, _hsl[2]/100, _hsl[3]/100);

            $hex = RGBtoHEX( _rgb.r, _rgb.g, _rgb.b );
        } else {
            console.error(value + ' is not a valid color');

            return null;
        }

        $rgb    = HEXtoRGB($hex);
        $hsl    = RGBtoHSL($rgb.r, $rgb.g, $rgb.b);
        $h      = Math.round($hsl.h * 360);
        $s      = Math.round($hsl.s * 100);
        $l      = Math.round($hsl.l * 100);

        __setValues(this, $rgb, $h, $s, $l);

        this.step = 1;
        this.__updateValues();

        return this;
    };

    Values.prototype.setStep = function setStep(value) {
        this.step = value;
        this.__updateValues();
        return this;
    };

    Values.prototype.getTints = function getTints(include_base) {
        var arr = this.tints;
        if (include_base) arr = arr.concat(this.__getCurrent());
        return arr;
    };

    Values.prototype.getShades = function getShades(include_base) {
        var arr = this.shades;
        if (include_base) arr = [this.__getCurrent()].concat(arr);
        return arr;
    };

    Values.prototype.getAll = function getAll() {
        return this.all;
    };

    Values.prototype.lightness = function lightness(value) {
        if ( typeof value === 'number') {
            var $h = this._hsl.h,
                $s = this._hsl.s,
                $l = this._hsl.l,
                $lightness = ($l + value);

            if ($lightness < 0)   lightness = 0;
            if ($lightness > 100) lightness = 100;

            var $rgb = HSLtoRGB($h/360, $s/100, $lightness/100);
            var $obj = __setValues({}, $rgb, $h, $s, $lightness);

            return $obj;
        }

        console.error('lightness expects a number');
    };

    Values.prototype.__updateValues = function __updateValues() {
        this.tints  = this.__getTints();
        this.shades = this.__getShades();
        this.all = this.tints.concat(this.__getCurrent(), this.shades);
    };

    Values.prototype.__getCurrent = function __getCurrent() {
        return __setValues({}, this._rgb, this._hsl.h, this._hsl.s, this._hsl.l);
    };

    Values.prototype.__getTints = function __getTints() {
        var $h = this._hsl.h,
            $s = this._hsl.s,
            $l = this._hsl.l,
            $i = 100,
            $tints  = [];

        while ($i > parseInt($l, 10)) {
            var $rgb = HSLtoRGB($h/360, $s/100, $i/100);
            $tints[$tints.length] = __setValues({}, $rgb, $h, $s, $i);
            $i -= this.step;
        };

        return $tints;
    };

    Values.prototype.__getShades = function __getShades() {
        var $h = this._hsl.h,
            $s = this._hsl.s,
            $l = this._hsl.l,
            $i = 0,
            $shades = [];

        while ($i < parseInt($l, 10)) {
            var $rgb = HSLtoRGB($h/360, $s/100, $i/100);
            $shades[$shades.length] = __setValues({}, $rgb, $h, $s, $i);
            $i += this.step;
        };

        return $shades.reverse();
    };

    if (typeof module !== 'undefined' && typeof module.exports !== 'undefined') {
        module.exports = Values;
    } else window.Values = Values;

})();
